design/005-Wire-Encoding-Abstraction.md

```diff
@@ -0,0 +1,108 @@
+`pkg/io/serialize.go`. (However, I just realize that `*big.Int`s are actually
+encoded in big-endian byte order...so we might as well switch to big-endianess
+in `pkg/io/serialize.go` to be more consistent
+([related issue](https://github.com/hyperledger-labs/go-perun/issues/167)).)
+
+## Rationale
+
+See #motivation.
+
+## Impact
+
+This can be implemented in a backwards-compatible fashion, where `pkg/io`
+encoding would be the default and other serialization backends could be injected
+at runtime.
+
+## Implementation
+
+Luckily, this can be achieved with a few changes to package `wire` and the
+`wire.Msg` implementations itself.
+
+The core idea is to create an envelope serialization backend, which can be set
+at runtime (using `init()` of the respecive serialization backend
+implementation). This backend's functions will then be used by
+`Envelope.(En|De)code` of package `wire`. The `pkg/io` serialization would then
+just be the default backend. Furthermore, the `Encode` methods of all `Msg`
+implementations would also be moved to this backend.
+
+Some of the proposed steps:
+* Create _Wire Serialization Backend_
+```go
+package wire
+
+// Backend is a serialization backend for wire messages. It should be set by a
+// backend's init() function.
+type EnvelopeSerializer interface {
+	func Encode(env *Envelope, w io.Writer) error
+	func Decode(r io.Reader) (*Envelope, error)
+}
+
+// current serializer
+var envelopeSerializer EnvelopeSerializer
+
+func EncodeEnvelope(env *Envelope, w io.Writer) error {
+	return envelopeSerializer.Encode(env, w)
+}
+// same for DecodeEnvelope
+
+// Allows to set the serialization backend at most once in an init function.
+// Panics if called twice.
+func SetEnvelopeSerializer(EnvelopeSerializer)
+```
+* `Envelope.(En|De)code` just call the functions of this backend, or we remove
+  those functions and use the free functions `(En|De)codeEnvelope`.
+
+* Current implementations of `Envelope.(En|De)code`, which are `pkg/io`
+  specific, will be moved to a new backend implementation package
+  `wire/perun`.
+  * Together with this move many other functions currently in package `wire`,
+    like `(En|De)code`, `DecodeAddress` and possibly many others that are
+    specific to `pkg/io` serialization.
+
+* We remove `perun.Encoder` from the `Msg interface`, Since en/decoding is now
+  injected via the backend.
+
+* We also move all implementations of methods `(En|De)code` on all `wire.Msg`
+  implementations, which currently are contained in package `client`, into
+  the new `wire/perun` package.
+  * This would also make the scope of the `client` package cleaner.
+  * I'd propose to register the encoders in a mapping similar to the `decoders`
+    in file `wire/msg.go`.
+
+Note that it is sufficient to create an abstraction of the envelope
+serialization, because, e.g., the `wire/net` package does serialization via the
+`Envelope`, see file `wire/net/ioconn.go`.
```

Comment:
Some more information could be added, where and how the encoding is used, i.e., where and how this would be injected and integrated within the current codebase. During our call, it was mentioned that the main entry point would be in `wire/net/ioconn.go`. How would this look like afterwrds? Any other usages?

Reply:
Besides network communication serialization, the only other place where this is used is persistence. There it would be nice to be able to choose the serialization protocol too, independent of the network serialization protocol. I'll add something about this.

Reply:
Added a section about persistence.